### ISSUE TITLE
 train mandarin  ,eval stop_token_loss increase

### ISSUE THREAD

**Issue by ChristopheZhao** [Rayhane-mamah/Tacotron-2#377]
hi,i try to train mandarin audio corpus use @begeekmyfriend 's griffin-lim branch code,alignment is good at step 10000，but synthesized wav(time long:08:14) is very bad expect the first 4 seconds, run demo_server.py on web,not matter how long the text input, it almost synthesize 37 seconds long audio file.and i found the eval stop_token_loss increased:
![image](https://user-images.githubusercontent.com/17317538/56779634-6c75c880-680e-11e9-8ab5-48265b580e88.png)

![image](https://user-images.githubusercontent.com/17317538/56779598-4e0fcd00-680e-11e9-9654-83500b56fe7c.png)

[synthesize_step10000.zip](https://github.com/Rayhane-mamah/Tacotron-2/files/3119505/synthesize_step10000.zip)

Anyone has same issues as me,and give me some advise,thank you!

**Comment by ChristopheZhao** [comment]
and my params : GTA = TRUE,step-10000-eval_waveform_linear.wav is clear and sounds good, but  synthesis is really bad.
[step-10000-eval-waveform-linear.zip](https://github.com/Rayhane-mamah/Tacotron-2/files/3119564/step-10000-eval-waveform-linear.zip)

**Comment by begeekmyfriend** [comment]
Please wait until the stop token loss reducing to zero. By the way, did you adapt the best ratio of N:T of the guided attention to your own corpus?

**Comment by ChristopheZhao** [comment]
@begeekmyfriend thanks for your response and code branch !! i use the biaobei dataset for train, and i set max_mel_frames = 600, max_text_length = 300,outputs_per_step=3 to calculate N and T. Is such parameter setting appropriate? if the stop token loss still increase,how can i do?and i change the mask_encoder = True in order to Speed up the convergence of the model ,  Does this change affect the training results?i'm not familiar with tts,thank you.

**Comment by begeekmyfriend** [comment]
Sorry I forgot to commit my newest hyper parameters for biaobei corpus. Please checkout https://github.com/begeekmyfriend/Tacotron-2/commit/6f5fe3f71d9ec592df5e003524f8d2bce7661c4d
By the way setting `mask_encoder` as `True` would not help convergence.

**Comment by ChristopheZhao** [comment]
@begeekmyfriend thank you, i will try.

**Comment by ChristopheZhao** [comment]
@begeekmyfriend ,thanks for your advise, i trained the model after modify the max_mel_frames = 600, max_text_length = 150,but it doesn't work after step 45000 training,it almost synthesize 38 seconds long audio after 10 Chinese word input,and the quality of synthesis audio is very bad.The train loss and stop_token_loss curve seems normal, eval is not good as train.
![image](https://user-images.githubusercontent.com/17317538/56873184-2ebdad80-6a63-11e9-92a2-281a71deec71.png)
train curve
![image](https://user-images.githubusercontent.com/17317538/56873202-45640480-6a63-11e9-81c6-081784aa461d.png)
eval curve
i'm confused about it, could you give me some advise about it ,thank you!

**Comment by begeekmyfriend** [comment]
Please be patient and wait until the stop token curve stay on the horizon axis. The results would be all right at that time. I promise. You might try synthesizing at 60K-step.
![image](https://user-images.githubusercontent.com/6031938/56873449-f5863d00-6a64-11e9-9edd-1db1b5c15f32.png)

**Comment by ChristopheZhao** [comment]
@begeekmyfriend ok, look forward good result after more step train, thanks again.

**Comment by begeekmyfriend** [comment]
By the way, in my test the encoder mask indicator has always been `False`. Is your attention (alignment) still kept?

**Comment by ChristopheZhao** [comment]
@begeekmyfriend em,i changed mask_encoder to True after i saw a annotation in code,"#DO NOT MAKE THIS BIGGER THAN 1 IF YOU DIDN'T TRAIN TACOTRON WITH "mask_encoder=True"!!",I don't quite understand the meaning of the parameter， it's better to set it to "False"?
![image](https://user-images.githubusercontent.com/17317538/56874102-169d5c80-6a6a-11e9-8610-66e0a8fcde6d.png)

**Comment by ChristopheZhao** [comment]
May i ask your more question?how about you eval stop_token_loss? My train stop_token_loss has been reduced during the train，but the eval has been increased so far .

**Comment by begeekmyfriend** [comment]
There might be some problems in the calculation of the eval loss. So do not care about this. Let us forcus on the effect of the evaluation results.
`mask_encoder` would enables `memory_seqence_length` in attention mechanism. See [here](https://github.com/begeekmyfriend/Tacotron-2/blob/mandarin-griffin-lim/tacotron/models/attention.py#L151). However without it we can as well synthesize out good results.

**Comment by ChristopheZhao** [comment]
ok，i see, Thank you for your patience.

**Comment by ChristopheZhao** [comment]
Hi,@begeekmyfriend ,After 110,000 steps of training, I synthesized 27 audios based on 27 Chinese Pinyin sentences, but only 3 pronunciations and pauses were good, 11 synthesized audios were 1 to 3 words less and echo problem, and another were poor audio echoes and word sequence confusion.  can you give me some advice, thank you again!

training_events:
![image](https://user-images.githubusercontent.com/17317538/57422582-46efb280-7243-11e9-8c26-386300449631.png)

good synthesized audio: 
("bai2 jia1 xuan1 hou4 lai2 yin2 yi3 hao2 zhuang4 de shi4 yi4 sheng1 li3 qu3 guo4 qi1 fang2 nv3 ren2 .")
[g_eval_batch_000.zip](https://github.com/Rayhane-mamah/Tacotron-2/files/3160066/g_eval_batch_000.zip)

words less and echo problem synthesized audio:
("na4 shi4 xi1 yuan2 shang4 gong3 jia1 cun1 da4 hu4 gong3 zeng1 rong2 de tou2 sheng1 nv3" )
[l2_eval_batch_002.zip](https://github.com/Rayhane-mamah/Tacotron-2/files/3160078/l2_eval_batch_002.zip)

word sequence confusion and echo problem synthesized audio:
[eval_batch_020.zip](https://github.com/Rayhane-mamah/Tacotron-2/files/3160090/eval_batch_020.zip)

**Comment by begeekmyfriend** [comment]
It is strange that the evaluation is not as good as expected while your training curves look good. What did the [target lengths](https://github.com/begeekmyfriend/Tacotron-2/blob/mandarin-griffin-lim/tacotron/synthesizer.py#L168) look like?

**Comment by begeekmyfriend** [comment]
By the way, there is no long sentences in Biaobei corpus, therefore you need to synthesize part of the long sentences and concatenate them all.
```shell
printf "file %s\n" *.wav > list.txt
ffmpeg -f concat -i list.txt -acodec copy eval_bailuyuan.wav
```

**Comment by ChristopheZhao** [comment]
> It is strange that the evaluation is not as good as expected while your training curves look good. What did the [target lengths](https://github.com/begeekmyfriend/Tacotron-2/blob/mandarin-griffin-lim/tacotron/synthesizer.py#L168) look like?

the tartget lengths like that when run synthesize.py with eval mode:
![image](https://user-images.githubusercontent.com/17317538/57440025-d021dc00-7279-11e9-844e-817a754f036d.png)
target lengths = 2000  synthesize 25 second long audio although it only took 3 or 5 seconds can synthesize.

**Comment by begeekmyfriend** [comment]
It is still strange that why there are so many `target lengths` with the same value 2000? Did you modify any code of my fork?

**Comment by ChristopheZhao** [comment]
i didn't modify hparams, but just modify the /dataset/preprocessor.py, because the biaobei dataset i download that doesn't has ‘trn’ file ,so i  modified it.

my biaobei dataset content structure:
![image](https://user-images.githubusercontent.com/17317538/57446997-97d5ca00-7288-11e9-8c3a-db3537c01025.png)


i don't use the .interval document.Read pinyin of each sentence from PhoneLableing/000001-010000.txt,
and get the corresponding wav file from Wave dir:
![image](https://user-images.githubusercontent.com/17317538/57447019-a328f580-7288-11e9-8286-fc0c84f8245e.png)

 
![image](https://user-images.githubusercontent.com/17317538/57447043-b045e480-7288-11e9-85bd-dd1cb3015ab2.png)


the length  of audio seems normal from generated train.txt:
![image](https://user-images.githubusercontent.com/17317538/57446888-534a2e80-7288-11e9-9447-024e45ab7997.png)
 

your code:
![image](https://user-images.githubusercontent.com/17317538/57445839-cbfbbb80-7285-11e9-9e43-d79fec289f9d.png)
my code:
![image](https://user-images.githubusercontent.com/17317538/57445885-e766c680-7285-11e9-9066-323c95927903.png)

Anything wrong with my code? Sorry, I am not familiar with TTS.

**Comment by begeekmyfriend** [comment]
The input text you have extracted from the corpus seems in binary format. Try converting them into string format please.

**Comment by ChristopheZhao** [comment]
@begeekmyfriend I didn't notice this error before， i will retrain model after modify this careless mistake，Thank you！

**Comment by begeekmyfriend** [comment]
I have fixed the failure of stop token prediction. Please merge this latest commit https://github.com/begeekmyfriend/Tacotron-2/commit/9401b414829b05efa542aaf3dfd055bfe49d5d8a

**Comment by begeekmyfriend** [comment]
By the way, the stop token prediction loss has reduced to zero and there should not be error in such token. Fortunately I have reappeared your issue. It is due to the input sentences that lead to such problems.
The evaluation should be all right if you input with this sentence.
> "bai2 jia1 xuan1 hou4 lai2 yin2 yi3 hao2 zhuang4 de shi4 yi4 sheng1 li3 qu3 guo4 qi1 fang2 nv3 ren2 ."

And such input is all right as well.
> "bai2 jia1 xuan1 hou4 lai2 yin2 yi3 hao2 zhuang4 de shi4 yi4 sheng1 li3 qu3 guo4 qi1 fang2 nv3 ren2"

But the inputs as follows would lead to such problem because the last character in this sentence is a whitespace that never seen in your training dataset. And the model has no idea whether to stop or not.
> "bai2 jia1 xuan1 hou4 lai2 yin2 yi3 hao2 zhuang4 de shi4 yi4 sheng1 li3 qu3 guo4 qi1 fang2 nv3 ren2 "
> "bai2 jia1 xuan1 hou4 lai2 yin2 yi3 hao2 zhuang4 de shi4 yi4 sheng1 li3 qu3 guo4 qi1 fang2 nv3 ren2. "

Anyway, you need to trim the end of each input text.

**Comment by ChristopheZhao** [comment]
@begeekmyfriend Thank you for your reply. I run your default branch, it has synthesized good quality audio after 30000 step training, really better than before!
"bai2 jia1 xuan1 hou4 lai2 yin2 yi3 wei2 ao4 de shi4 yi4 sheng1 li3 qu3 guo4 qi1 fang2 nv3 ren2."
[eval_batch_000.zip](https://github.com/Rayhane-mamah/Tacotron-2/files/3187013/eval_batch_000.zip)

"na4 shi4 xi1 yuan2 shang4 gong3 jia1 cun1 da4 hu4 gong3 zeng1 rong2 de tou2 sheng1 nv3,"
[eval_batch_002.zip](https://github.com/Rayhane-mamah/Tacotron-2/files/3187014/eval_batch_002.zip)
Only two little problem,It's hard to listened clearly about the pronunciation of 'de',another is some pause in sentence split word to two char pronunciation ，the  word pronunciation is not continuous ，especially in longer sentence. Such as , "wo xiang yao qu wai mian kan kan", the synthesize audio is good, but one bad pause in "wo xiang yao qu wai mian kan geng mei de feng jing",and missing the 'de' pronunciation too. Did think that reason is traning data is not enough？
[geng mei de feng jing.zip](https://github.com/Rayhane-mamah/Tacotron-2/files/3187080/geng.mei.de.feng.jing.zip)

[wo xiang yao qu wai mian kan kan.zip](https://github.com/Rayhane-mamah/Tacotron-2/files/3187087/wo.xiang.yao.qu.wai.mian.kan.kan.zip)


[wo xiang yap qu wai mian kan geng mei de feng jing.zip](https://github.com/Rayhane-mamah/Tacotron-2/files/3187083/wo.xiang.yap.qu.wai.mian.kan.geng.mei.de.feng.jing.zip)

**Comment by begeekmyfriend** [comment]
Forgot to tell you that as for Biaobei transcipt you need 'de5' and 'le5' rather than 'de' and 'le'.

**Comment by HaiFengZeng** [comment]
@begeekmyfriend,Hi! I'm facing problem of overfitting and the eval stop token loss is increasing, do you meet the same issue?

**Comment by begeekmyfriend** [comment]
I would not be responsible for any bad code modification.

**Comment by ChristopheZhao** [comment]
may i ask your more question?i check the eval step in training,eval data equal training data multiplily a test ratio ,the eval sentence format is like that:
"fang3 zhen1 zhi1 cu1 hua1 ni2 deng3 yin4 hua1 mian4 liao4 de5 qiao3 miao4 yun4 yong4",i don't find the space white in sentence, so why the eval stop token loss still increase while training?

**Comment by ChristopheZhao** [comment]
@begeekmyfriend OK, i got it ,modify 'de' to 'de5' ,because almost 'de' was notationed 'de5'  in the training data ,thank you!

**Comment by chen849157649** [comment]
> brother, @begeekmyfriend , the training cuve,   the stop token curve stay on the horizon axis. as follow,
> 
![image](https://user-images.githubusercontent.com/32058220/58525656-22ae4280-81ff-11e9-8581-6df28aadabea.png),
but , eval cuve is that, 
![image](https://user-images.githubusercontent.com/32058220/58525735-6143fd00-81ff-11e9-9050-c08d38f403ac.png),  
i  Use mandarin-griffin-lim branch, set linear_prediction as False, then begin train biaobei corpus, 
could you give me some advise about it ,thank you!

**Comment by begeekmyfriend** [comment]
No need to care about eval curves, they do nothing with evaluation results.

**Comment by chen849157649** [comment]
@begeekmyfriend ,ok, thank you , sample of synthesize griffin_lim with mel , it sounds quite good,  thank you share,

**Comment by begeekmyfriend** [comment]
You might use your GTA mel evaluation to train [WaveRNN](https://github.com/fatchord/WaveRNN) for better results.

**Comment by nanma** [comment]
> No need to care about eval curves, they do nothing with evaluation results.

Hi @begeekmyfriend, I'm wondering why do you say the eval curves do nothing with evaluation results？ I check the source code, and the eval loss is calculated with 5% of the training data. Is there something wrong with the evaluation process?

**Comment by begeekmyfriend** [comment]
I do not look closely into the code yet. Please look at the evaluation stop token loss, does it reasonable?

**Comment by nanma** [comment]
@begeekmyfriend Yes, sometimes the eval loss is indeed not reasonable, I'm not sure for the reason.
